File: hs/day4.hs
import Control.Exception(assert)
import Data.ByteString(ByteString)
import qualified Data.ByteString as BS(readFile)
import qualified Data.ByteString.Char8 as C
import Data.Word

data Range = Range {
    start, end :: Word8
}

rangeLength :: Range -> Word8
rangeLength a = end a - start a

rangeContains :: Range -> Range -> Bool
rangeContains a b = start big <= start small && end big >= end small
    where
        (big, small) = if rangeLength a >= rangeLength b then (a, b) else (b, a)

rangeOverlaps :: Range -> Range -> Bool
rangeOverlaps a b = start a  <= end b && end a >= start b

toPairs :: ByteString -> (Range, Range)
toPairs line = assert (length parsePair == 2) (elf1, elf2)
    where
        parsePair = parseRange <$> C.split ',' line
        parseElems r = read . C.unpack <$> C.split '-' r :: [Word8]
        assembleRange rl = assert (length rl == 2) $ Range (head rl) (last rl)
        parseRange = assembleRange . parseElems
        elf1 = head parsePair
        elf2 = last parsePair

part1 :: [(Range, Range)] -> Int
part1 = length . filter (uncurry rangeContains)

part2 :: [(Range, Range)] -> Int
part2 = length . filter (uncurry rangeOverlaps)

main :: IO ()
main = do
    input <- BS.readFile "../input/2022/day4.txt"
    let pairs = toPairs <$> C.lines input
    print $ part1 pairs
    print $ part2 pairs

File: src/day4.rs
use aoc_runner_derive::{aoc, aoc_generator};

use std::ops::RangeInclusive;

type RangePair = (RangeInclusive<u8>, RangeInclusive<u8>);

#[aoc_generator(day4)]
fn pairs(input: &str) -> Vec<RangePair> {
    input
        .lines()
        .map(|l| {
            let ranges = l
                .split(',')
                .map(|r| {
                    let nums = r.split('-').collect::<Vec<_>>();
                    RangeInclusive::new(nums[0].parse().unwrap(), nums[1].parse().unwrap())
                })
                .collect::<Vec<_>>();
            (ranges[0].clone(), ranges[1].clone())
        })
        .collect()
}

#[aoc(day4, part1)]
fn part1(input: &[RangePair]) -> usize {
    input
        .iter()
        .filter(|pairs| {
            let (smaller, bigger) = if pairs.0.len() <= pairs.1.len() {
                (&pairs.0, &pairs.1)
            } else {
                (&pairs.1, &pairs.0)
            };
            bigger.start() <= smaller.start() && bigger.end() >= smaller.end()
        })
        // .map(|m| {println!("Overlap: {:?}", m); m})
        .count()
}

#[aoc(day4, part2)]
fn part2(input: &[RangePair]) -> usize {
    input
        .iter()
        .filter(|pairs| pairs.0.start() <= pairs.1.end() && pairs.0.end() >= pairs.1.start())
        .count()
}
